04-graphs/05-application.md

```diff
@@ -0,0 +1,704 @@
+*Fig. Representing a maze as (a) a grid, (b) a collection of nodes, and as (c) an adjacency list. A lot of the code that programmers write relates to converting among various data representations.*  
+
+<br />
+
+<details style="max-width: 700px; margin: auto;">
+  <summary>Click here to see the adjacency list source as text.</summary>
+
+```py
+graph = {
+    (0, 0): [(0, 1)],
+    (0, 1): [(0, 0), (1, 1)],
+    (1, 1): [(0, 1), (2, 1)],
+    (2, 1): [(1, 1), (2, 2)],
+    (2, 2): [(2, 1)],
+}
+```
+
+</details>
+
+There are a few ways we could process the grid representation. Assuming we write a function `convert_maze_to_graph` that takes a maze grid `maze` as input and returns an adjacency list of nodes, each node represented by a tuple of its row and column, and the list of adjacent nodes, think about what the steps to convert the maze to a graph might be. Then take a look at our steps!
+
+<br />
+
+<details style="max-width: 700px; margin: auto;">
+  <summary>Click here to see our steps!</summary>
+
+1. Create an empty dictionary to hold the adjacency information.
+2. Iterate over each row in the maze.
+3. Iterate over each column in the row.
+4. If the current cell is a wall, skip it, since the path cannot pass through a wall.
+5. Otherwise, create a tuple representing the current cell `(r, c)`.
+6. Create an empty list to hold the adjacent nodes.
+7. For each of the four directions (up, right, down, left):
+    1. Calculate the row and column of the adjacent cell.
+    2. If the adjacent cell is out of bounds or a wall, skip it.
+    3. Otherwise, create a tuple representing the adjacent cell.
+    4. Add the tuple to the list of adjacent nodes.
+8. Add the tuple and list of adjacent nodes to the dictionary.
+9. Return the dictionary.
+
+</details>
+
+![A grid laid over the top-left corner marked (A) shows that the spaces to the left and above are invalid (outside the maze representation), while the spaces to the right and below are valid. A grid laid over a space along the top edge marked (B) shows that only the space above is invalid.](images/graphs_application_maze_corner_edge.png)  
+*Fig. (A) has no location above or to the left in our maze representation. (B) has no location above it. Locations in the other corners and along the other edges have similar restrictions. We must inspect all of the neighboring locations, looking for a passage or a wall, and use that information to build a graph of the connected passage areas.*
```

Comment:
That's a fair point. I do want to show that we need to be careful about which cells we try to access, but since this image shows up before the text about the complication, I could see a student tying it back to the "building a passage graph" mention from earlier.

I think what I'm going to do is split the following paragraph so that the reader sees the complication text before the image, anf the helper method part would come after. I'll also update the caption to make it more specically about checking a valid cell at all rather than checking for a passage or a wall.